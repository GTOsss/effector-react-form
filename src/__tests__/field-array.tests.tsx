import {Controller} from '../../index';
import React from 'react';
import {useForm, useFieldArray} from '../index';
import {render, screen, fireEvent} from '@testing-library/react';
import {createStore} from 'effector';

const getId = (() => {
  let counter = 0;

  return () => counter += 1;
})();

const $values = createStore({});
const $fieldsInline = createStore({});

interface InputProps {
  controller: Controller,
  label: string,
}

const validateRequired = (value) => value ? undefined : 'Field is required';

const Input: React.FC<InputProps> = ({
  controller,
  label,
}) => {
  const {input, fieldState, form, error} = controller();

  const isShowError = error && (form.submitted || form.hasOuterError || fieldState.blurred);

  return (
    <div role="wrapper-for-input" className="input-wrap">
      <label>{label}</label>
      <input
        {...input}
        value={input.value || ''}
        role="textbox"
        className={`input${isShowError ? ' input-error' : ''}`}
        placeholder={label}
      />
      {isShowError && (<span>{error}</span>)}
    </div>
  );
};

const FormItems = ({controller, name, onPush}) => {
  const {map, push, remove} = useFieldArray({name, $values, $fieldsInline});

  return (
    <div className="formsItem" role="formItem">
      {map(({fieldName, index, field}) => (
        <div key={field.id} className="formItem">
          <Input
            label="Username"
            controller={controller({name: `${fieldName}.username`, validate: validateRequired})}
          />
          <Input
            label="First name"
            controller={controller({name: `${fieldName}.profile.firstName`, validate: validateRequired})}
          />
          <button type="button" onClick={() => remove(index)}>
            remove form item
          </button>
        </div>
      ))}
      <button type="button" onClick={() => onPush(push)}>
        add form item with values
      </button>
    </div>
  );
};

const FieldArray = ({onPush}) => {
  const {handleSubmit, controller} = useForm({
    onSubmit: () => null,
    $values,
    $fieldsInline,
  });

  return (
    <form onSubmit={handleSubmit}>
      <FormItems name="parents" controller={controller} onPush={onPush} />
      <button type="submit">submit</button>
    </form>
  );
};

describe('FieldArray', () => {
  test('render', () => {
    render(<FieldArray onPush={() => null} />);
    const formItems = screen.getAllByRole('formItem');
    expect(formItems).toMatchSnapshot();
  });

  test('add form item with values > render', () => {
    render(<FieldArray onPush={(push) => push({id: getId(), username: 'gto', profile: {firstName: 'Timofey'}})} />);
    const addFormItem = screen.getByText('add form item with values');
    fireEvent.click(addFormItem);
    const formItems = screen.getAllByRole('formItem');
    expect(formItems).toMatchSnapshot();
  });

  test('add form item with values > $values', () => {
    render(<FieldArray onPush={(push) => push({id: getId(), username: 'gto', profile: {firstName: 'Timofey'}})} />);
    const addFormItem = screen.getByText('add form item with values');
    fireEvent.click(addFormItem);
    expect($values.getState()).toMatchSnapshot();
  });

  test('add form item with values (x3) > $values', () => {
    const formItems = [
      {id: getId(), username: '0'},
      {id: getId(), username: '1'},
      {id: getId(), username: '2'},
    ];

    let counter = 0;
    const onPush = (push) => {
      push(formItems[counter++]);
    };
    render(<FieldArray onPush={onPush} />);
    const addFormItem = screen.getByText('add form item with values');
    fireEvent.click(addFormItem);
    fireEvent.click(addFormItem);
    fireEvent.click(addFormItem);
    expect($values.getState()).toMatchSnapshot();
  });

  test('add form item with values (x3) > remove(1) > $values', () => {
    const formItems = [
      {id: getId(), username: '0'},
      {id: getId(), username: '1'},
      {id: getId(), username: '2'},
    ];

    let counter = 0;
    const onPush = (push) => {
      push(formItems[counter++]);
    };
    render(<FieldArray onPush={onPush} />);
    const addFormItem = screen.getByText('add form item with values');
    fireEvent.click(addFormItem);
    fireEvent.click(addFormItem);
    fireEvent.click(addFormItem);
    const removeFormItem = screen.getAllByText('remove form item');
    fireEvent.click(removeFormItem[1]);
    expect($values.getState()).toMatchSnapshot();
  });

  test('add form item with values (x3) > remove(1) > $fieldsInline', () => {
    const formItems = [
      {id: getId(), username: '0'},
      {id: getId(), username: '1'},
      {id: getId(), username: '2'},
    ];

    let counter = 0;
    const onPush = (push) => {
      push(formItems[counter++]);
    };
    render(<FieldArray onPush={onPush} />);
    const addFormItem = screen.getByText('add form item with values');
    fireEvent.click(addFormItem);
    fireEvent.click(addFormItem);
    fireEvent.click(addFormItem);
    const removeFormItem = screen.getAllByText('remove form item');
    fireEvent.click(removeFormItem[1]);
    expect($fieldsInline.getState()).toMatchSnapshot();
  });
});
