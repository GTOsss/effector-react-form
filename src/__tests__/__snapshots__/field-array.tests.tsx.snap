// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FieldArray add form item with values (x3) > $values 1`] = `
Object {
  "parents": Array [
    Object {
      "id": 1,
      "username": "0",
    },
    Object {
      "id": 2,
      "username": "1",
    },
    Object {
      "id": 3,
      "username": "2",
    },
  ],
}
`;

exports[`FieldArray add form item with values (x3) > remove(1) > $fieldsInline 1`] = `
Object {
  "parents.0.profile.firstName": Object {
    "_type": "fieldMeta",
    "active": false,
    "blurred": false,
    "blurredAfterOuterError": false,
    "changed": false,
    "changedAfterOuterError": false,
    "touched": false,
    "touchedAfterOuterError": false,
    "validate": [Function],
  },
  "parents.0.username": Object {
    "_type": "fieldMeta",
    "active": false,
    "blurred": false,
    "blurredAfterOuterError": false,
    "changed": false,
    "changedAfterOuterError": false,
    "touched": false,
    "touchedAfterOuterError": false,
    "validate": [Function],
  },
  "parents.1.profile.firstName": Object {
    "_type": "fieldMeta",
    "active": false,
    "blurred": false,
    "blurredAfterOuterError": false,
    "changed": false,
    "changedAfterOuterError": false,
    "touched": false,
    "touchedAfterOuterError": false,
    "validate": [Function],
  },
  "parents.1.username": Object {
    "_type": "fieldMeta",
    "active": false,
    "blurred": false,
    "blurredAfterOuterError": false,
    "changed": false,
    "changedAfterOuterError": false,
    "touched": false,
    "touchedAfterOuterError": false,
    "validate": [Function],
  },
}
`;

exports[`FieldArray add form item with values (x3) > remove(1) > $values 1`] = `
Object {
  "parents": Array [
    Object {
      "id": 1,
      "username": "0",
    },
    Object {
      "id": 3,
      "username": "2",
    },
  ],
}
`;

exports[`FieldArray add form item with values > render 1`] = `
Array [
  <div
    class="formsItem"
    role="formItem"
  >
    <div
      class="formItem"
    >
      <div
        class="input-wrap"
        role="wrapper-for-input"
      >
        <label>
          Username
        </label>
        <input
          class="input"
          name="parents.0.username"
          placeholder="Username"
          role="textbox"
          value="gto"
        />
      </div>
      <div
        class="input-wrap"
        role="wrapper-for-input"
      >
        <label>
          First name
        </label>
        <input
          class="input"
          name="parents.0.profile.firstName"
          placeholder="First name"
          role="textbox"
          value="Timofey"
        />
      </div>
      <button
        type="button"
      >
        remove form item
      </button>
    </div>
    <button
      type="button"
    >
      add form item with values
    </button>
  </div>,
]
`;

exports[`FieldArray render 1`] = `
Array [
  <div
    class="formsItem"
    role="formItem"
  >
    <button
      type="button"
    >
      add form item with values
    </button>
  </div>,
]
`;
